opc/resource_database_service_instance.go

```diff
@@ -0,0 +1,594 @@
+	if err != nil {
+		return fmt.Errorf("Error creating DatabaseServiceInstance: %s", err)
+	}
+
+	d.SetId(info.Name)
+	return resourceOPCDatabaseServiceInstanceRead(d, meta)
+}
+
+func resourceOPCDatabaseServiceInstanceRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] Resource state: %#v", d.State())
+	client := meta.(*OPCClient).databaseClient.ServiceInstanceClient()
+	if client == nil {
+		return fmt.Errorf("Database Client is not initialized. Make sure to use `database_endpoint` variable or `OPC_DATABASE_ENDPOINT` env variable")
+	}
+
+	log.Printf("[DEBUG] Reading state of ip reservation %s", d.Id())
+	getInput := database.GetServiceInstanceInput{
+		Name: d.Id(),
+	}
+
+	result, err := client.GetServiceInstance(&getInput)
+	if err != nil {
+		// DatabaseServiceInstance does not exist
+		if opcClient.WasNotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading database service instance %s: %s", d.Id(), err)
+	}
+
+	if result == nil {
+		d.SetId("")
+		return nil
+	}
+
+	log.Printf("[DEBUG] Read state of database service instance %s: %#v", d.Id(), result)
+	d.Set("name", result.Name)
+	d.Set("description", result.Description)
+	d.Set("apex_url", result.ApexURL)
+	d.Set("backup_destination", result.BackupDestination)
+	d.Set("backup_supported_version", result.BackupSupportedVersion)
+	d.Set("char_set", result.CharSet)
+	d.Set("cloud_storage_container", result.CloudStorageContainer)
+	d.Set("compute_site_name", result.ComputeSiteName)
+	d.Set("connect_descriptor", result.ConnectDescriptor)
+	d.Set("connect_descriptor_with_public_ip", result.ConnectorDescriptorWithPublicIP)
+	d.Set("created_by", result.CreatedBy)
+	d.Set("creation_time", result.CreationTime)
+	d.Set("current_version", result.CurrentVersion)
+	d.Set("dbaasmonitor_url", result.DBAASMonitorURL)
+	d.Set("edition", result.Edition)
+	d.Set("em_url", result.EMURL)
+	d.Set("failover_database", result.FailoverDatabase)
+	d.Set("glassfish_url", result.GlassFishURL)
+	d.Set("hdg_prem_ip", result.HDGPremIP)
+	d.Set("hybrid_db", result.HybridDG)
+	d.Set("identity_domain", result.IdentityDomain)
+	d.Set("ip_network", result.IPNetwork)
+	d.Set("ip_reservations", result.IPReservations)
+	d.Set("jaas_instances_using_service", result.JAASInstancesUsingService)
+	d.Set("level", result.Level)
+	d.Set("listener_port", result.ListenerPort)
+	d.Set("n_char_set", result.NCharSet)
+	d.Set("num_ip_reservations", result.NumIPReservations)
+	d.Set("num_nodes", result.NumNodes)
+	d.Set("pdb_name", result.PDBName)
+	d.Set("rac_database", result.PDBName)
+	d.Set("region", result.Region)
+	d.Set("uri", result.URI)
+	d.Set("shape", result.Shape)
+	d.Set("sid", result.SID)
+	d.Set("sm_plugin_version", result.SMPluginVersion)
+	d.Set("subscription_type", result.SubscriptionType)
+	d.Set("timezone", result.Timezone)
+	d.Set("total_shared_storage", result.TotalSharedStorage)
+	d.Set("version", result.Version)
+	return nil
```

Comment:
We should add a TODO here to set the Parameter and vm_ssh_key fields when available